### PR TITLE
Boiler plate code of HttpSink for #874.

### LINKED ISSUE #874: Provide an HTTP Sink plugin
**Is your feature request related to a problem? Please describe.**

Data Prepper can output to OpenSearch, but cannot output to arbitrary HTTP endpoints.

**Describe the solution you'd like**

Create a new sink in Data Prepper which outputs JSON to an arbitrary HTTP endpoint.

It can output events to an HTTP endpoint as a JSON array of events.

Example:

```
[{"key1": "value1", "key2": "value2"}, {"key1": "value3", "key2": "value4"}]
```

It could also be configured to write each new event as a JSON string on its own line. This is similar to the OpenSearch bulk API.

Example:

```
{"key1": "value1", "key2": "value2"}
{"key1": "value3", "key2": "value4"}
```

**Additional Context**

This sink should be able to write to the existing `http` source plugin and its existing [JsonCodec](https://github.com/opensearch-project/data-prepper/blob/main/data-prepper-plugins/http-source/src/main/java/org/opensearch/dataprepper/plugins/source/loghttp/codec/JsonCodec.java).

**Possible user configurations**

* `url` - The URL to send events to
* `proxy` - A proxy server to send requests through
* `http_method` - Which HTTP method to use. Defaults to `POST`.
* `format` - Configuration to determine the format of the data. May be `json-array` for a JSON array, or `json-nd` for JSON-ND.
* Pluggable authentication options. Should support HTTP Basic, Bearer token, and mTLS to start
* Configurations to specify the SSL certificates or chains to trust (File, S3 object, ACM). This can be similar to the options provided in the `opensearch` sink.
* SSL verification modes similar to [core peer forwarding](https://github.com/opensearch-project/data-prepper/blob/7a6f747aab3ebc7f2ab731fa6a1c8c9d5efaa3bb/data-prepper-core/src/main/java/org/opensearch/dataprepper/peerforwarder/PeerForwarderConfiguration.java#L71-L72).
* Options to control retrying failed requests (which status codes; how many retries; backoff approach)
* Timeout options (connect, request, socket)

### PR DESCRIPTION
**Pull request by mallikagogoi7**
### Description
Boiler plate code of HttpSink for #874.
 
### Issues Resolved
Resolves #874
 
### Check List
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [X] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: data-prepper-plugins/http-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/AwsServiceClientFactory.java [hunk 1]
```diff
@@ -0,0 +1,13 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package org.opensearch.dataprepper.plugins.sink;
+
+/**
+ * AWS factory class required to create AWS Http Endpoint client for this plugin.
```
[the remark this answers is not in the document]
Comment (mallikagogoi7) [reply data-prepper-plugins/http-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/AwsServiceClientFactory.java:8]: Thanks for the Review @kkondaka, 
This class is to get the AWS service connections, Example: if we want to test AWS hosted Http Endpoint(SageMaker, Lambda etc)

File: data-prepper-plugins/http-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/accumulator/BufferFactory.java [hunk 2]
```diff
@@ -0,0 +1,10 @@
+/*
```
[the remark this answers is not in the document]
Comment (mallikagogoi7) [reply data-prepper-plugins/http-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/accumulator/BufferFactory.java:1]: Thanks for the review @kkondaka, 
@ashoktelukuntla @dlvenable can you please suggest on this requirement.

[the remark this answers is not in the document]
Comment (ashoktelukuntla) [reply data-prepper-plugins/http-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/accumulator/BufferFactory.java:1]: I would vote for common code , if code is same as S3 Sink. @mallikagogoi7 - do you have any specific / tailored implementation to HTTP Sink ?

[the remark this answers is not in the document]
Comment (mallikagogoi7) [reply data-prepper-plugins/http-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/accumulator/BufferFactory.java:1]: @ashoktelukuntla, Thanks for the confirmation. I shall move the code to common package and update the PR

[the remark this answers is not in the document]
Comment (mallikagogoi7) [reply data-prepper-plugins/http-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/accumulator/BufferFactory.java]: As per your suggestion, I have moved the code to common place, please review.

File: data-prepper-plugins/common/src/main/java/org/opensearch/dataprepper/plugins/accumulator/Buffer.java [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    /**
+     * Gets the current size of the buffer. This should be the number of bytes.
+     * @return buffer size.
+     */
+    long getSize();
+    int getEventCount();
+
+    long getDuration();
+
+    void flushToS3(S3Client s3Client, String bucket, String key) ;
```
Comment (oeyh) [inline data-prepper-plugins/common/src/main/java/org/opensearch/dataprepper/plugins/accumulator/Buffer.java:26]: This is specific to S3 sink, not something common. I think you will still need `sendDataToHttpEndpoint` for http sink.

Comment (mallikagogoi7) [reply data-prepper-plugins/common/src/main/java/org/opensearch/dataprepper/plugins/accumulator/Buffer.java:27]: Thanks for the review @oeyh , Added sendDataToHttpEndpoint() in Buffer

Comment (mallikagogoi7) [reply data-prepper-plugins/common/src/main/java/org/opensearch/dataprepper/plugins/accumulator/Buffer.java]: Thanks for the review @oeyh, We have removed the methods and shall implement in the respective plugins.

**Review by oeyh — APPROVED** [review]

**Review by kkondaka — APPROVED** [review]

**Pull request merged**